ZA | 25-SDC-JULY | Luke Manyamazi | Sprint 1 | New Feature Rebloom

Luke-Manyamazi opened the pull request:
<!--

You must title your PR like this:

Region | Cohort | FirstName LastName | Sprint | Assignment Title

For example,

London | 25-ITP-May | Carol Owen | Sprint 1 | Alarm Clock

Fill in the template below - remove any sections that don't apply.

Complete the self checklist - replace each empty box in the checklist [ ] with a [x].

Add the label "Needs Review" and you will get review.

Respond to volunteer reviews until the volunteer marks it as "Complete".

-->

## Learners, PR Template

Self checklist

- [x] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
- [x] My changes meet the requirements of the task
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)

## Changelist

1. Add a "re-bloom" button so users can share a bloom to their own feed and followers.
2. UI should clearly show original author and who re-bloomed it, distinguishing it from the user’s own blooms.
3. Display rebloom count on each bloom; timestamp can reflect rebloom time to indicate when it was shared.

## Questions

Ask any questions you have for your reviewer.

illicitonion commented on backend/data/blooms.py:
```diff
@@ -4,37 +4,203 @@
+            # Keep error logging for debugging purposes
+            print(f"Error adding bloom: {e}")
+            # Returning 0 or raising an error are typical ways to indicate failure
+            return 0
```
I don't think any other code in this codebases uses magic return values to indicate errors, and the code that calls this doesn't pass this error back to the client. You need to handle errors differently here.

Luke-Manyamazi replied:
Good point — using magic return values hides failures from the caller.
I’ve refactored the rebloom logic to raise exceptions instead.

illicitonion commented on backend/data/blooms.py:
```diff
@@ -4,37 +4,203 @@
+
+    if sender_user is None:
+        print(f"Error adding rebloom: User with ID {user_id} not found.")
+        return None
```
This doesn't bubble the error back up to the caller - the caller will need to know this failed.

This applies throughout your change.

illicitonion commented on backend/endpoints.py:
```diff
@@ -178,6 +196,89 @@ def get_bloom(id_str):
+        })
+
+@jwt_required()
+def check_rebloom_status(bloom_id_str):
```
What are the trade-offs of having this be a stand-alone function that's called per bloom vs including this information up-front when getting a list of blooms?

illicitonion commented on backend/data/blooms.py:
```diff
@@ -4,37 +4,203 @@
+        try:
+            # First create the new bloom
+            rebloom_bloom_id = add_bloom(
+                sender=sender_user,  # <<< Use the fully hydrated User object
```
AFAICT `add_bloom` only actually uses the sender ID, an no other data from the sender - might it be simpler to make `add_bloom` take just the id, and change the other call-sites from `sender` to `sender.id`, rather than needing to add a whole extra database roundtrip to turn the sender ID into a sender, just to take only the ID back out of it?

illicitonion commented on backend/data/users.py:
```diff
@@ -106,3 +106,27 @@ def generate_salt() -> bytes:
+    This is necessary for operations like reblooming where only the ID is known.
+    """
+    with db_cursor() as cur:
+        # NOTE: Adjust column names if they are different in your 'users' table
```
What does this comment mean?

illicitonion commented on backend/endpoints.py:
```diff
@@ -149,6 +149,24 @@ def do_follow():
     )
 
+@jwt_required()
+def do_unfollow():
```
This looks unrelated to this change?

illicitonion commented on backend/endpoints.py:
```diff
@@ -178,6 +196,89 @@ def get_bloom(id_str):
+@jwt_required(optional=True)
+def rebloom_bloom(bloom_id_str):
+    print(f"🔍 Rebloom endpoint called with: {bloom_id_str}")
+    if request.method == "OPTIONS":
```
None of the other endpoints handle OPTIONS requests explicitly - why does this one need to?

illicitonion commented on backend/endpoints.py:
```diff
@@ -178,6 +196,89 @@ def get_bloom(id_str):
+
+    if existing_rebloom:
+        # Undo rebloom: delete the rebloom from user's timeline
+        success = blooms.delete_rebloom(existing_rebloom['id'])
```
It's surprising that re-blooming would delete an existing bloom. None of the other endpoints have this toggle behaviour, e.g. `/follow` doesn't unfollow if called on an existing follow.

illicitonion commented on backend/endpoints.py:
```diff
@@ -178,6 +196,89 @@ def get_bloom(id_str):
+        rebloom_id = blooms.add_rebloom(
+            user_id=current_user.id,
+            original_bloom_id=bloom_id,
+            content=original_bloom.content  # FIX: .content not ['content']
```
What is this comment about?

illicitonion commented on backend/main.py:
```diff
@@ -29,18 +32,21 @@ def main():
     app.json = CustomJsonProvider(app)
 
-    # Configure CORS to handle preflight requests
+    # --- CORS FIX APPLIED HERE ---
```
Things seemed to be working ok without this change - why did you need to change it?

illicitonion commented on db/schema.sql:
```diff
@@ -26,3 +26,27 @@ CREATE TABLE hashtags (
 );
+
+-- Add rebloom_count to blooms table (to track total reblooms)
+ALTER TABLE blooms ADD COLUMN rebloom_count INTEGER DEFAULT 0;
```
Ideally you wouldn't need to track rebloom counts in the database - there is a race condition here where values can get out of date (e.g. if you create a rebloom but haven't updated this count yet).

Instead, use SQL relationships to compute this - in a query, you should be able to compute the count of blooms which have this ID as an original_bloom_id

Your Python and JS code for querying a Bloom shouldn't have to change, but it removes the need for you to update the original bloom from the python side.

illicitonion commented on db/schema.sql:
```diff
@@ -26,3 +26,27 @@ CREATE TABLE hashtags (
+ALTER TABLE blooms ADD COLUMN original_bloom_id BIGINT REFERENCES blooms(id);
+
+-- Create reblooms table to track the rebloom relationship
+CREATE TABLE reblooms (
```
This table shouldn't need to exist - all of the information should already be present in the blooms table for you to do some joins to infer this information.

illicitonion replied:
As far as I can tell this table isn't used any more, but it's still created? Please remove.

illicitonion commented on front-end/components/bloom.mjs:
```diff
@@ -31,57 +24,134 @@ const createBloom = (template, bloom) => {
+
+  const rebloomIndicator = document.createElement('div');
+  rebloomIndicator.className = 'rebloom-indicator';
+  rebloomIndicator.innerHTML = `
```
There's an XSS vulnerability here by using `innerHTML`

Luke-Manyamazi commented:
Hi Daniel

I have done the following:

1. Fixed rebloom creation by fetching full User objects and preventing missing argument errors.
2. Removed rebloom_count updates in Python; counts are now computed dynamically via SQL to avoid race conditions.
3. Secured frontend against XSS by replacing innerHTML with safe DOM methods and properly encoding hashtags and usernames.
4. Improved endpoint error handling, type checks, and JWT integration for login, registration, and rebloom actions.

illicitonion commented on backend/data/users.py:
This file seems to have a lot of unrelated changes to it?

illicitonion commented on backend/endpoints.py:
```diff
@@ -178,6 +197,67 @@ def get_bloom(id_str):
+
+    already_rebloomed = blooms.has_user_rebloomed(current_user.id, bloom_id)
+
+    if already_rebloomed:
```
This kind of "toggle" behaviour can be complicated to support well in clients - they may retry requests (e.g. if the network blipped) which can accidentally undo changes.

In general I'd recommend API design where a user clearly expresses intent to rebloom or un-rebloom something, and will get an error if that's not possible, rather than sometimes undoing the action. The UI can re-use the same button, but it should typically talk to a different backend endpoint.

illicitonion reviewed (commented):
A lot of this is looking a lot better, but there's still a lot of unrelated changes in here, and a bunch of unused code. Please take a look, and make sure:
1. You've addressed and resolved all comments.
2. All of the code in this PR is needed and related to this change.

It's much harder for a reviewer to see if your code is working or look for issues when half of the change isn't related.

Luke-Manyamazi commented:
Mistakenly closed this PR

Luke-Manyamazi commented:
Opening it to finish with review edits

illicitonion commented on backend/endpoints.py:
```diff
@@ -8,240 +16,199 @@
 
 MINIMUM_PASSWORD_LENGTH = 5
 
+# -------- Helper Functions ---------
```
It's quite hard to review your change with all of these unrelated changes

illicitonion reviewed (commented):
This generally looks decent, but it doesn't look like it works for me?

I ran it locally, and when I first logged in I got an error prompt.
I dismissed it and tried to rebloom something, and got an error alert" `Unexpected token '<', "<!DOCTYPE "... is not valid JSON`

Luke-Manyamazi commented:
"Hi Daniel, I’ve refactored the Rebloom feature to address your feedback:

Fixed XSS: Replaced innerHTML with createTextNode for safe rendering.

Removed Race Conditions: rebloom_count is now calculated via SQL subqueries instead of a manual column.

JSON Fix: Standardized all error responses to jsonify() to resolve the Unexpected token < crash.

Efficiency: Streamlined add_bloom to accept sender_id directly, removing redundant DB lookups.

Luke-Manyamazi commented:
Hi Daniel, I was getting an error while trying to rebloom, but it's fixed now. 

Here’s a summary of the changes:

Fix: Rebloom Functionality & Auth Persistence
I resolved the 404/422 errors by aligning the frontend and backend layers:
Database: Added the missing rebloom_id column to the blooms table.
Routing: Corrected the fetch URL in bloom.mjs (removed the /api prefix) to match the Flask route.
Auth: Fixed the "Token: null" issue by ensuring apiService saves the JWT to localStorage during login/signup.
The feature now successfully creates reblooms with a valid 200 OK response.
